diff --git a/src/hci/tui/settings_ui.c b/src/hci/tui/settings_ui.c
index be421cc..d0f5b71 100644
--- a/src/hci/tui/settings_ui.c
+++ b/src/hci/tui/settings_ui.c
@@ -384,10 +384,10 @@ static void draw_instruction_row ( struct settings_ui *ui ) {
 	if ( ui->row.editing ) {
 		msg ( INSTRUCTION_ROW,
 		      "Enter - accept changes" INSTRUCTION_PAD
-		      "Ctrl-C - discard changes" );
+		      "ESC/Ctrl-C - discard changes" );
 	} else {
 		msg ( INSTRUCTION_ROW,
-		      "%sCtrl-X - exit configuration utility",
+		      "%sESC/Ctrl-X - exit configuration utility",
 		      ( ( ui->row.origin == ui->settings ) ?
 			"Ctrl-D - delete setting" INSTRUCTION_PAD : "" ) );
 	}
@@ -486,6 +486,7 @@ static int main_loop ( struct settings *settings ) {
 				if ( ( rc = save_setting ( &ui ) ) != 0 )
 					alert ( " %s ", strerror ( rc ) );
 				/* Fall through */
+			case ESC:
 			case CTRL_C:
 				select_setting_row ( &ui, ui.scroll.current );
 				redraw = 1;
@@ -526,6 +527,7 @@ static int main_loop ( struct settings *settings ) {
 			select_setting_row ( &ui, ui.scroll.current );
 			redraw = 1;
 			break;
+		case ESC:
 		case CTRL_X:
 			return 0;
 		case CR:
